File: app/src/main/java/ch/unstable/ost/views/StopDotView.java
package ch.unstable.ost.views;


import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.Canvas;
import android.graphics.Paint;
import androidx.annotation.Nullable;
import android.util.AttributeSet;
import android.view.View;

import ch.unstable.ost.R;


public class StopDotView extends View {

    private final Paint mCirclePaint;
    private final int mCirclePixelWidth;
    private final Paint mLinePaint;
    private Type mType;

    public StopDotView(Context context, @Nullable AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public StopDotView(Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        TypedArray style = context.getTheme().obtainStyledAttributes(attrs, R.styleable.StopDotView, defStyleAttr, 0);
        int lineColor = style.getColor(R.styleable.StopDotView_lineColor, 0xff000000);
        int lineWidthPx = style.getDimensionPixelSize(R.styleable.StopDotView_lineWidth, 1);
        int type = style.getInt(R.styleable.StopDotView_lineMode, 0);
        mCirclePixelWidth = style.getDimensionPixelSize(R.styleable.StopDotView_circleWidth, 2);
        int circleColor = style.getColor(R.styleable.StopDotView_circleColor, 0xffffffff);
        switch (type) {
            case 1:
                mType = Type.TOP;
                break;
            case 2:
                mType = Type.BOTTOM;
                break;
            default:
                mType = Type.BOTH;
        }
        mLinePaint = new Paint(Paint.ANTI_ALIAS_FLAG);
        mLinePaint.setColor(lineColor);
        mLinePaint.setStrokeWidth(lineWidthPx);
        mLinePaint.setStyle(Paint.Style.STROKE);       // set to STOKE
        mLinePaint.setStrokeJoin(Paint.Join.ROUND);    // set the join to round you want
        mLinePaint.setStrokeCap(Paint.Cap.ROUND);      // set the paint cap to round too
        mCirclePaint = new Paint(Paint.ANTI_ALIAS_FLAG);
        mCirclePaint.setColor(circleColor);
        mCirclePaint.setStyle(Paint.Style.FILL);
        style.recycle();
    }

    public void setLineMode(Type type) {
        if(mType != type) {
            mType = type;
            invalidate();
        }
    }

    @Override
    protected void onDraw(Canvas canvas) {
        super.onDraw(canvas);
        float x = getWidth() / 2f;
        float startY = 0;
        float stopY = getHeight();
        switch (mType) {
            case TOP:
                startY = getHeight() / 2f;
                break;
            case BOTH:
                break;
            case BOTTOM:
                stopY = getHeight() / 2f;
                break;
        }
        canvas.drawLine(x, startY, x, stopY, mLinePaint);
        canvas.drawCircle(x, getHeight() / 2f, mCirclePixelWidth / 2f, mCirclePaint);
    }

    protected Paint getLinePaint() {
        return mLinePaint;
    }

    protected Paint getCirclePaint() {
        return mCirclePaint;
    }

    protected int getCirclePixelWidth() {
        return mCirclePixelWidth;
    }

    public enum Type {
        TOP,
        BOTTOM,
        BOTH
    }
}

File: app/src/main/java/ch/unstable/ost/views/StopDotView.kt
package ch.unstable.ost.views

import android.content.Context
import android.graphics.Canvas
import android.graphics.Paint
import android.util.AttributeSet
import android.view.View
import ch.unstable.ost.R

class StopDotView @JvmOverloads constructor(context: Context,
                                            attrs: AttributeSet? = null,
                                            defStyleAttr: Int = 0,
                                            defStyleRes: Int = 0) : View(context, attrs, defStyleAttr, defStyleRes) {
    private val circlePaint: Paint
    private val circlePixelWidth: Int
    private val linePaint: Paint
    var lineMode: LineMode = LineMode.BOTH
        set(value) {
            if (value != field) {
                field = value
                invalidate()
            }
        }

    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)
        val x = width / 2f
        var startY = 0f
        var stopY = height.toFloat()
        when (lineMode) {
            LineMode.TOP -> startY = height / 2f
            LineMode.BOTH -> {
            }
            LineMode.BOTTOM -> stopY = height / 2f
        }
        canvas.drawLine(x, startY, x, stopY, linePaint)
        canvas.drawCircle(x, height / 2f, circlePixelWidth / 2f, circlePaint)
    }

    enum class LineMode {
        TOP, BOTTOM, BOTH
    }

    init {
        val style = context.theme.obtainStyledAttributes(attrs, R.styleable.StopDotView, defStyleAttr, 0)
        val lineColor = style.getColor(R.styleable.StopDotView_lineColor, -0x1000000)
        val lineWidthPx = style.getDimensionPixelSize(R.styleable.StopDotView_lineWidth, 1)
        val type = style.getInt(R.styleable.StopDotView_lineMode, 0)
        circlePixelWidth = style.getDimensionPixelSize(R.styleable.StopDotView_circleWidth, 2)
        val circleColor = style.getColor(R.styleable.StopDotView_circleColor, -0x1)
        lineMode = when (type) {
            1 -> LineMode.TOP
            2 -> LineMode.BOTTOM
            else -> LineMode.BOTH
        }
        linePaint = Paint(Paint.ANTI_ALIAS_FLAG)
        linePaint.color = lineColor
        linePaint.strokeWidth = lineWidthPx.toFloat()
        linePaint.style = Paint.Style.STROKE // set to STOKE
        linePaint.strokeJoin = Paint.Join.ROUND // set the join to round you want
        linePaint.strokeCap = Paint.Cap.ROUND // set the paint cap to round too
        circlePaint = Paint(Paint.ANTI_ALIAS_FLAG)
        circlePaint.color = circleColor
        circlePaint.style = Paint.Style.FILL
        style.recycle()
    }
}